src/games/strategy/triplea/ai/proAI/ProAI.java

```diff
@@ -482,11 +490,11 @@ public boolean selectAttackSubs(final Territory unitTerritory) {
     final List<Unit> defenders = (List<Unit>) battle.getDefendingUnits();
     ProLogger.info(player.getName() + " checking sub attack in " + unitTerritory + ", attackers=" + attackers
         + ", defenders=" + defenders);
-    ProBattleUtils.setData(getGameData());
+    calc.setData(getGameData());
 
     // Calculate battle results
     final ProBattleResult result =
-        ProBattleUtils.calculateBattleResults(player, unitTerritory, attackers, defenders, new HashSet<Unit>(), true);
+        calc.calculateBattleResults(player, unitTerritory, attackers, defenders, new HashSet<Unit>(), true);
```

Comment:
What if this method took a gamedata object? Then you wouldn't have to worry about tracking state in the calc object. You could get rid of the setData call on 493.